kubectl-cloudflow/deploy/deploy.go

```diff
@@ -92,6 +86,328 @@ func GetCloudflowApplicationDescriptorFromDockerImage(dockerRegistryURL string,
+
+func makeDeployment(bp blueprint, streamletID string, streamlet domain.Streamlet,
+	deployImages map[string]string, replicas map[string]int) domain.Deployment {
+
+	var deployment domain.Deployment
+	deployment.StreamletName = streamletID
+	deployment.ClassName = streamlet.Descriptor.ClassName
+	image, err := bp.getImageFromStreamletID(streamletID)
+	if err != nil {
+		util.LogAndExit(err.Error())
+	}
+	deployment.Image = deployImages[image]
+	deployment.PortMappings = getAllPortMappings(bp.name, streamletID, streamlet.Descriptor, bp)
+	deployment.VolumeMounts = streamlet.Descriptor.VolumeMounts
+	deployment.Runtime = streamlet.Descriptor.Runtime
+	deployment.SecretName = transformToDNS1123SubDomain(streamlet.Name)
+	deployment.Name = fmt.Sprintf("%s.%s", bp.name, streamletID)
+	if scale, ok := replicas[streamletID]; ok {
+		deployment.Replicas = &scale
+	}
+
+	config, endpoint := getServerConfigAndEndpoint(bp.name, streamlet)
+	if config != nil {
+		deployment.Config = config
+	}
+	if endpoint != (domain.Endpoint{}) {
+		deployment.Endpoint = &endpoint
+	}
+	return deployment
+}
+
+// Removes from the leading and trailing positions the specified characters.
+func trim(name string) string {
+	return strings.TrimSuffix(strings.TrimSuffix(strings.TrimPrefix(strings.TrimPrefix(name, "."), "-"), "."), "-")
+}
+
+// Make a name compatible with DNS 1123 Subdomain
+// Limit the resulting name to 253 characters
+func transformToDNS1123SubDomain(name string) string {
+	splits := strings.Split(name, ".")
+	var normalized []string
+	for _, split := range splits {
+		normalized = append(normalized, trim(normalize(split)))
+	}
+	joined := strings.Join(normalized[:], ".")
+	if len(joined) > subDomainMaxLength {
+		return strings.TrimSuffix(joined[0:subDomainMaxLength], ".")
+	}
+	return strings.TrimSuffix(joined, ".")
+}
+
+func normalize(name string) string {
+	t := transform.Chain(norm.NFKD)
+	normalizedName, _, _ := transform.String(t, name)
+	s := strings.ReplaceAll(strings.ReplaceAll(strings.ToLower(normalizedName), "_", "-"), ".", "-")
+	var re = regexp.MustCompile(`[^-a-z0-9]`)
+	return re.ReplaceAllString(s, "")
+}
+
+func getServerAttribute(descriptor domain.Descriptor) domain.Attribute {
+	var attr domain.Attribute
+	for _, attribute := range descriptor.Attributes {
+		if attribute.AttributeName == "server" {
+			attr = attribute
+			return attribute
+		}
+	}
+	return attr
+}
+
+// MinimumEndpointContainerPort indicates the minimum value of the end point port
+const endpointContainerPort = 3000
+const subDomainMaxLength = 253
+
+type configRoot struct {
+	Cloudflow internal `json:"cloudflow,omitempty"`
+}
+type internal struct {
+	Internal server `json:"internal,omitempty"`
+}
+type server struct {
+	Server containerPort `json:"server,omitempty"`
+}
+type containerPort struct {
+	ContainerPort int `json:"container-port,omitempty"`
+}
+
+func getServerConfigAndEndpoint(appID string, streamlet domain.Streamlet) (json.RawMessage, domain.Endpoint) {
```

Comment:
Let's add proper documentation to all our high level functions. It improves code readability and its part of Go's way of doing things (https://blog.golang.org/godoc).

Reply:
This is private though? let's make a decision on code improvement doc.

Reply:
I was thinking that we should document only the _public_ functions. Should we also document private ones ? @skonto @RayRoestenburg